### ISSUE TITLE
DeferredLock documentation doesn't make FIFO-like nature clear

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3082]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)| @mwhudson reported|
|-|-|
|Trac ID|trac#3082|
|Type|defect|
|Created|2008-03-16 21:27:18Z|
|Branch|https://github.com/twisted/twisted/tree/deferredlock-fifo-3082|

Suppose .acquire() is called twice for a locked DeferredLock.  Which deferred gets called back when the lock is released?  Looking at the implementation, its pretty clear that its the first one returned, but the documentation doesn't mention this.

I assume it's intentional (DeferredLock would be useless for what I want it for if it wasn't :) and so it should be documented.

<details><summary>Searchable metadata</summary>

```
trac-id__3082 3082
type__defect defect
reporter__mwh mwh
priority__normal normal
milestone__ 
branch__branches_deferredlock_fifo_3082 branches/deferredlock-fifo-3082
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1205702838000000 1205702838000000
changetime__1266362910000000 1266362910000000
version__None None
owner__ 
cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

(In [25360]) Branching to 'deferredlock-fifo-3082'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

If someone can summarize this then I'll branch and add it to the docs, I'm not familiar with DeferredLock.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I think the ticket summary covers the idea pretty well.  The point is that the lock is given out in a FIFO manner.  The alternative, of course, is that it could be given out in LIFO order, or randomly, or something like that.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @khorn<br>@thijstriemstra set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_6"></a>|@khorn commented|
|-|-|

I like the way MochiKit describes this: [http://www.mochikit.com/doc/html/MochiKit/Async.html]

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_7"></a>|@khorn commented|
|-|-|

(In [28207]) Update DeferredLock docstring.

Update docstrings for DeferredLock.acquire() and DeferredLock.release() to better describe FIFO-like behavior

Refs: [#3082](https://github.com/twisted/twisted/issues/3082)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)|@khorn removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_9"></a>|@mwhudson commented|
|-|-|

There seems to be conflicts in http://twistedmatrix.com:12435/3082, but I do like the wording.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_10"></a>|@khorn commented|
|-|-|

(In [28208]) Resolve header conflict

Refs: [#3082](https://github.com/twisted/twisted/issues/3082)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess set owner to @khorn|
|-|-|

Thanks for the documentation work, khorn.

Since we're going for explicitness about being a FIFO, in acquire I'd say "the Deferred goes at the end of a waiting list" instead of "the Deferred goes into a waiting list".

Stylistically, while this whole file could use a lot of work (and thanks for shoring up some of the surrounding epytext), sticking to DeferredLock:
 * add an @rtype on acquire's return value
 * starting triple quote on a line by itself for acquire
 * not that the surrounding code is being good about this, but for DeferredLock at least, 2 lines between methods, 3 lines between classes
 * no need for escaping quotes in the DeferredLock class docstring

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_15"></a>|@khorn commented|
|-|-|

(In [28227]) Address review comments

 * tweaked DeferredLock.acquire docstring
 * stylistic fixes (whitespace, formatting, etc.)

Refs: [#3082](https://github.com/twisted/twisted/issues/3082)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_16"></a>|@khorn removed owner|
|-|-|

Addressed review comments, and added some whitespace to conform to coding standard while I was in there.

Back to review...

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_17"></a>|@jesstess set owner to @khorn|
|-|-|

Looks ready to merge.

Thanks for the work!

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_18"></a>|@khorn set status to `closed`|
|-|-|

(In [28229]) Merge deferredlock-fifo-3082: Improve DeferredLock Documentation

Improve docstring for twisted.internet.defer.DeferredLock to better
describe it's FIFO-like nature.  Also a bunch of whitespace/formatting
improvements.

Author: thijs, khorn
Reviewer: mwh, jesstess
Fixes: [#3082](https://github.com/twisted/twisted/issues/3082)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
